src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.SingleTC.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    using static BinaryOperatorKind;
+
+    internal static partial class ValueSetFactory
+    {
+        private struct SingleTC : FloatingTC<float>, NumericTC<float>
+        {
+            float NumericTC<float>.MinValue => float.MinValue;
+
+            float NumericTC<float>.MaxValue => float.MaxValue;
+
+            float FloatingTC<float>.NaN => float.NaN;
+
+            float FloatingTC<float>.MinusInf => float.NegativeInfinity;
+
+            float FloatingTC<float>.PlusInf => float.PositiveInfinity;
+
+            float NumericTC<float>.Next(float value)
+            {
+                Debug.Assert(!float.IsNaN(value));
+                Debug.Assert(!float.IsInfinity(value));
+                Debug.Assert(value != float.MaxValue);
+                if (value < 0)
+                {
+                    if (value == -float.Epsilon)
+                        return 0.0f; // skip negative zero
+                    return -UintAsFloat(FloatAsUint(-value) - 1);
+                }
+                return UintAsFloat(FloatAsUint(value) + 1);
+            }
+
+            private unsafe static uint FloatAsUint(float d)
+            {
+                float* dp = &d;
+                uint* lp = (uint*)dp;
+                return *lp;
+            }
+
+            private unsafe static float UintAsFloat(uint l)
+            {
+                uint* lp = &l;
+                float* dp = (float*)lp;
+                return *dp;
+            }
+
+            (float leftMax, float rightMin) NumericTC<float>.Partition(float min, float max)
+            {
+                Debug.Assert(min != max);
+
+                if (min == float.MinValue && max == float.MaxValue)
+                    return (-UintAsFloat(1), 0.0f); // skip negative zero
+
+                Debug.Assert((min >= 0) == (max >= 0));
+                if (min < 0)
+                {
+                    uint minl = FloatAsUint(-max);
+                    uint maxl = FloatAsUint(-min);
+                    uint midl = minl + (maxl - minl) / 2;
```

Comment:
>minl + (maxl - minl) / 2; [](start = 32, length = 25)

consider comment: this does not produce the float average, but does produce a partition of the float space between min and max
Maybe worth explaining why not just do the average between min and max